### PR TITLE
[gasless] Add remaining balance verification for withdrawals

### PR DESCRIPTION
**Pull request by alex-mysten**
## Summary

- Adds `gasless_verify_remaining_balance` protocol config flag (enabled in v122)
- At voting time, validates that a gasless withdrawal does not leave the sender's accumulator balance in an invalid state (remaining must be 0 or >= min_transfer)
- At execution time, validates the same invariant against actual accumulator net balance changes, covering partial redemptions and `withdrawal_split` flows
- Transactional tests covering boundary conditions for both validation paths

## Test plan

- [x] `cargo nextest run -p sui-adapter-transactional-tests gasless` (8 tests pass)
- [x] `cargo xclippy -D warnings`
- [x] Protocol config snapshots updated

### CODE REVIEW

File: crates/sui-core/src/accumulators/funds_read.rs [hunk 1]
```diff
@@ -49,4 +50,41 @@ pub trait AccountFundsRead: Send + Sync {
...
+                Balance::maybe_get_balance_type_param(type_tag).unwrap_or_else(|| type_tag.clone());
+            if let Some(&min_amount) = min_amounts.get(&coin_type)
+                && min_amount > 0
+                && remaining < min_amount as u128
+            {
+                return Err(SuiErrorKind::UserInputError {
+                    error: UserInputError::InvalidWithdrawReservation {
+                        error: format!(
+                            "Gasless withdrawal from {} leaves {remaining} which is \
+                             below minimum {min_amount} for token type {coin_type}",
```
Comment (aschran) [inline crates/sui-core/src/accumulators/funds_read.rs:79]: maybe say "below non-zero minimum" so it's a bit clearer that users may zero out their balances?

**Review by aschran — APPROVED** [review]

**Comment by mystenmark** [comment]
Looks like this has the odd property if your balance is `R + A + A` where A is the minimum balance, you could reserve `R +  A` but would then be unable to redeem only `R + A - 1`.

I do not think this is a problem than anyone will run into and its too hard to fix, but just wanted to make sure i'm understanding this correctly.

File: sui-execution/latest/sui-adapter/src/temporary_store.rs [hunk 2]
```diff
@@ -618,6 +622,26 @@ impl<'backing> TemporaryStore<'backing> {
             }
         }
 
+        if let Some(reservations) = withdrawal_reservations {
```
Comment (tnowacki) [inline sui-execution/latest/sui-adapter/src/temporary_store.rs:625]: Why do we need the withdrawal information here? Can't we just check the net_totals and filter out those that aren't for the sender?

File: crates/sui-core/src/accumulators/funds_read.rs (hunk 1 above)
Comment (tnowacki) [reply crates/sui-core/src/accumulators/funds_read.rs:79]: +1, I read this like three times and checked the test and was still confused lol. 
Maybe something like "Invalid gasless withdrawal from {}. Gasless transactions must either use the entire balance, or leave at least {min_amount} for token type {coin_type}, and the remaining amount of {remaining} is below that threshold" 
A bit wordy still, but trying to spell out it is either 0 or above the threshold

File: sui-execution/latest/sui-adapter/src/temporary_store.rs [hunk 3]
```diff
@@ -569,7 +569,11 @@ impl<'backing> TemporaryStore<'backing> {
     /// - No new objects were created or existing objects mutated (written_objects is empty)
     /// - The set of deleted objects exactly equals the set of input Coin objects
     /// - Each recipient receives at least the minimum transfer amount per token type
-    pub fn check_gasless_execution_requirements(&self) -> Result<(), String> {
+    /// - Unused withdrawal reservation (reservation - actual split) is 0 or >= min_amount
+    pub fn check_gasless_execution_requirements(
+        &self,
```
Comment (tnowacki) [inline sui-execution/latest/sui-adapter/src/temporary_store.rs:574]: Also, can we check this up front before execution? Obviously we cannot in general, but the current form of gasless is restrictive enough that we maybe could? (it could be at least approximate it with failing to allow execution when it cannot be statically determined)

Comment (alex-mysten) [reply sui-execution/latest/sui-adapter/src/temporary_store.rs:574]: This is exactly where I started - statically analyzing the PTB based on arguments and move commands at voting time to decide how much of the balance was consumed.

At voting time, none of the ptb structure or arguments have been semantically validated. There's a lot of validation surface area that would be moved up from execution to voting time - argument arity, type, overflow considerations, etc etc that we need to consider.

So the answer is yes, we can do this, but I'm not sure we want to be in the business of performing move argument validation upstream of execution. Also it would be brittle. If we have a feature requirement where the amount to send is computed, the static-validation model breaks.

File: sui-execution/latest/sui-adapter/src/temporary_store.rs (hunk 2 above)
Comment (alex-mysten) [reply sui-execution/latest/sui-adapter/src/temporary_store.rs:625]: We have two independent checks we want to apply:
1. at voting time, sender is prevented from sending a Withdrawal that would leave only dust in their balance.
2. at execution time, of that Withdrawal, they may partially or fully consume it, but they must not leave dust

net totals would not be sufficient as it knows nothing about the sender's actual balance

**Comment by alex-mysten** [comment]
> Looks like this has the odd property if your balance is `R + A + A` where A is the minimum balance, you could reserve `R + A` but would then be unable to redeem only `R + A - 1`.
> 
> I do not think this is a problem than anyone will run into and its too hard to fix, but just wanted to make sure i'm understanding this correctly.

@mystenmark correct. a ptb that leaves only dust remaining from the amount that they withdrew would be rejected. a small advantage of going the static analysis route is that we'd just compare against user balance at submission, so this case wouldn't happen. but this has the tradeoffs I called out above

File: sui-execution/latest/sui-adapter/src/temporary_store.rs (hunk 3 above)
Comment (tnowacki) [reply sui-execution/latest/sui-adapter/src/temporary_store.rs:574]: Yeah I agree on the brittle ness. I mostly saw it as a spam protection system, not so much as a way of actually validating the "correctness" of the txn. 

I think things would be different if we thought that this was all the expressivity we would have under gasless transactions--now and forever. But since the scope will increase, it is not the spot we should rely on.

File: crates/sui-adapter-transactional-tests/tests/gasless/gasless_remaining_balance.move [hunk 4]
```diff
@@ -0,0 +1,100 @@
+// Copyright (c) Mysten Labs, Inc.
```
Comment (tnowacki) [inline crates/sui-adapter-transactional-tests/tests/gasless/gasless_remaining_balance.move:1]: As a small nit, could you split these into failing and non-failing transactions? It makes it a lot easier for me to understand what behavior is allowed vs disallowed.

File: sui-execution/latest/sui-adapter/src/execution_engine.rs [hunk 5]
```diff
@@ -683,6 +691,37 @@ mod checked {
...
+            if let CallArg::FundsWithdrawal(fw) = input
+                && let Some(coin_type) = fw.type_arg.get_balance_type_param()
+            {
+                let owner = match fw.withdraw_from {
+                    WithdrawFrom::Sender => sender,
+                    WithdrawFrom::Sponsor => {
+                        debug_fatal!(
+                            "WithdrawFrom::Sponsor is not allowed in gasless transactions"
+                        );
+                        return None;
```
Comment (tnowacki) [inline sui-execution/latest/sui-adapter/src/execution_engine.rs:714]:
```suggestion
                        continue;
```

Comment (tnowacki) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs:714]: We probably want to harden here by still attempting to look at all the reservations

File: sui-execution/latest/sui-adapter/src/execution_engine.rs [hunk 6]
```diff
@@ -683,6 +691,37 @@ mod checked {
...
+    ) -> Option<BTreeMap<(SuiAddress, TypeTag), u64>> {
+        let TransactionKind::ProgrammableTransaction(pt) = transaction_kind else {
+            debug_fatal!("Gasless transaction must be a ProgrammableTransaction");
+            return None;
+        };
+        let sender = tx_ctx.sender();
+        let mut reservations = BTreeMap::<(SuiAddress, TypeTag), u64>::new();
+        for input in &pt.inputs {
+            if let CallArg::FundsWithdrawal(fw) = input
+                && let Some(coin_type) = fw.type_arg.get_balance_type_param()
```
Comment (tnowacki) [inline sui-execution/latest/sui-adapter/src/execution_engine.rs:706]: I feel we probably shouldn't silently fail right now when this is none? Maybe? 
In short, I think it is impossible for this to be anything other than a balance at the moment so it should maybe be like 

```suggestion
           let CallArg::FundsWithdrawal(fw) = input else { continue };
           let Some(coin_type) = fw.type_arg.get_balance_type_param() else {
               debug_fatal!("we should only have balance types for withdrawals at this point");
               continue;
            };
```

File: sui-execution/latest/sui-adapter/src/temporary_store.rs (hunk 2 above)
Comment (tnowacki) [reply sui-execution/latest/sui-adapter/src/temporary_store.rs:625]: Do we have a minimum transfer amount? Or just a minimum balance remaining

File: sui-execution/latest/sui-adapter/src/temporary_store.rs [hunk 7]
```diff
@@ -618,6 +622,26 @@ impl<'backing> TemporaryStore<'backing> {
...
 
+        if let Some(reservations) = withdrawal_reservations {
+            for ((owner, token_type), &reserved) in reservations {
+                let net = net_totals
+                    .get(&(*owner, token_type.clone()))
+                    .copied()
+                    .unwrap_or(0);
+                let remaining = (reserved as i128).saturating_add(net);
+                if remaining > 0
+                    && let Some(&min_amount) = allowed_types.get(token_type)
```
Comment (tnowacki) [inline sui-execution/latest/sui-adapter/src/temporary_store.rs:633]:
```suggestion
                    && let Some(&min_balance_remaining) = allowed_types.get(token_type)
```

Saying `min_balance_remaining` instead of `min_amount` might have made it easier for me to grok this at the first go around (worth changing in the other spot too maybe)

File: sui-execution/latest/sui-adapter/src/execution_engine.rs [hunk 8]
```diff
@@ -683,6 +691,37 @@ mod checked {
...
+                    WithdrawFrom::Sender => sender,
+                    WithdrawFrom::Sponsor => {
+                        debug_fatal!(
+                            "WithdrawFrom::Sponsor is not allowed in gasless transactions"
+                        );
+                        return None;
+                    }
+                };
+                let Reservation::MaxAmountU64(amount) = fw.reservation;
+                let entry = reservations.entry((owner, coin_type)).or_default();
```
Comment (tnowacki) [inline sui-execution/latest/sui-adapter/src/execution_engine.rs:718]: Nit only for my general distrust of `Default` (you also don't use `unwrap_or_default()` below so maybe just consistency 🤷 take or leave it, I don't care really. We just hate `Default` in external crates land)
```suggestion
                let entry = reservations.entry((owner, coin_type)).or_insert(0);
```

File: sui-execution/latest/sui-adapter/src/temporary_store.rs (hunk 2 above)
Comment (alex-mysten) [reply sui-execution/latest/sui-adapter/src/temporary_store.rs:625]: We have both. Minimum transfer amount is already implemented

File: crates/sui-core/src/accumulators/funds_read.rs (hunk 1 above)
Comment (alex-mysten) [reply crates/sui-core/src/accumulators/funds_read.rs]: Addressed

File: crates/sui-adapter-transactional-tests/tests/gasless/gasless_remaining_balance_accept.move [hunk 9]
```diff
@@ -0,0 +1,100 @@
+// Copyright (c) Mysten Labs, Inc.
```
Comment (alex-mysten) [reply crates/sui-adapter-transactional-tests/tests/gasless/gasless_remaining_balance_accept.move:1]: Split these as suggested

File: sui-execution/latest/sui-adapter/src/execution_engine.rs (hunk 5 above)
Comment (alex-mysten) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs]: We now look at all the reservations. I left the debug_assert since we would probably want to check this again if we did ever change the rule to disallow sponsor reservations

File: sui-execution/latest/sui-adapter/src/execution_engine.rs (hunk 6 above)
Comment (alex-mysten) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs]: Addressed

File: sui-execution/latest/sui-adapter/src/temporary_store.rs (hunk 7 above)
Comment (alex-mysten) [reply sui-execution/latest/sui-adapter/src/temporary_store.rs]: Addressed

File: sui-execution/latest/sui-adapter/src/execution_engine.rs (hunk 8 above)
Comment (alex-mysten) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs]: Addressed

**Review by tnowacki — APPROVED** [review]
